## Data Science / Data Analyst Popular Method
* 1.Introduction to pandas
* 2.Top 5 or bottom 5 records
* 3.renaming the column names
* 4.statistical description
* 5.adding new column
* 6.dropping column
* 7.Setting any column as index
* 8.resettig index
* 9.selecting row with index position
* 10.selecting subsection of the dataset with LOC and iloc
* 11.how to drop the row
* 12.conditional filtering using "&" and "|"
* 13.apply function on single column
* 14.apply func on multiple column using lambda func
* 15.sorting method
* 16.min , max and their index position
* 17.value_counts/unique/nunique/replace/map function
* 18.treatment of duplicate values
* 19.nlargest/nsmallest to get highest/lowest records
* 20.sample of the dataset (by numbers or percentage)
* 21.handling missing data
* 22.isnull/notnull
* 23.dropping missing values
* 24.filling missing values
* 25.group by operation on pandas
* 26.combining dataframe-->concatenation
* 27.combining dataframe--> merging-->join--> inner/left/right/outer
* 28.text method on string data
* 29.cleaning the data


### 1 Introduction to Pandas 
-->>>Pandas is the open source python libraries which is used for EDA , Data cleaning. 
* pandas dataframe it is two dimentional tabular data structure with labelled axis (rows and columns) 
* Dataframe is table of columns and rows in pandas that we can easily restructure and filter
* Formal Defination:=A group of pandas series Object that share the same index

### 2 Top 5 or bottom 5 records

In [2]:
### We have imported the pandas library and the seaborn as well seaborn we import for dummy dataset 
import pandas as pd
import seaborn as sns

In [3]:
### Here's the list of Dummy DataFrame which is alreaday present in seaborn dataframe 
sns.get_dataset_names()

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'dowjones',
 'exercise',
 'flights',
 'fmri',
 'geyser',
 'glue',
 'healthexp',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'seaice',
 'taxis',
 'tips',
 'titanic']

In [305]:
### So we are gonna to use titanic dataframe
df = sns.load_dataset('titanic')
print("--------------------------->>> TOP 5 VALUE <<<------------------------","\n")
display(df.head(5))
print()
print("--------------------------->>> Bottom 5 VALUE <<<------------------------","\n")
df.tail(5)

--------------------------->>> TOP 5 VALUE <<<------------------------ 



survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True


--------------------------->>> Bottom 5 VALUE <<<------------------------ 



survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

### 3. renaming the column names

In [93]:
### If we want to change the column in that particular dataframe we must need to use the permeter inplace 
df.rename(columns ={"sex":"Gender"}, inplace = True)

In [38]:
df

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 4 .statistical description

In [40]:
df.describe().T

count       mean        std   min      25%      50%   75%       max
survived  891.0   0.383838   0.486592  0.00   0.0000   0.0000   1.0    1.0000
pclass    891.0   2.308642   0.836071  1.00   2.0000   3.0000   3.0    3.0000
age       714.0  29.699118  14.526497  0.42  20.1250  28.0000  38.0   80.0000
sibsp     891.0   0.523008   1.102743  0.00   0.0000   0.0000   1.0    8.0000
parch     891.0   0.381594   0.806057  0.00   0.0000   0.0000   0.0    6.0000
fare      891.0  32.204208  49.693429  0.00   7.9104  14.4542  31.0  512.3292

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   Gender       891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


### 5.adding new column

In [47]:
df["Temp Col"] = df['sibsp'] * df['fare']
df

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  Temp Col  
0      man        True  NaN  Southampton    no  False    7.2500  
1    woman       False    C    Cherbourg   yes  False   71.2833  
2    woman       False  NaN  Southampton   yes   True    0.0000  
3    woman       False    C  Southampton   yes  False   53.1000  
4      man        True  NaN  Southampton    no   True    0.0000  
..     ...         ...  ...          ...   ...    ...       ...  
886    man        True  NaN  Southampton    no   True    0.0000  
887  woman       False    B  Southampton   yes   True    0.0000  
888  woman       False  NaN  Southampton    no  False   23.4500  
889    man        True    C    Cherbourg   yes   True    0.0000  
890    man        True  NaN   Queenstown    no   True    0.0000  

[891 rows x 16 columns]

### .dropping column

In [49]:
df.drop('Temp Col' , inplace = True , axis = 1)
df

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 7.Setting any column as index

In [56]:
df.set_index('alone', inplace = True)
df

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
alone                                                                           
False         0       3    male  22.0      1      0   7.2500        S   Third   
False         1       1  female  38.0      1      0  71.2833        C   First   
True          1       3  female  26.0      0      0   7.9250        S   Third   
False         1       1  female  35.0      1      0  53.1000        S   First   
True          0       3    male  35.0      0      0   8.0500        S   Third   
...         ...     ...     ...   ...    ...    ...      ...      ...     ...   
True          0       2    male  27.0      0      0  13.0000        S  Second   
True          1       1  female  19.0      0      0  30.0000        S   First   
False         0       3  female   NaN      1      2  23.4500        S   Third   
True          1       1    male  26.0      0      0  30.0000        C   First   
True          0       3    male  32.0      0      0   7.7500        Q   Third   

         who  adult_male deck  embark_town alive  
alone                                             
False    man        True  NaN  Southampton    no  
False  woman       False    C    Cherbourg   yes  
True   woman       False  NaN  Southampton   yes  
False  woman       False    C  Southampton   yes  
True     man        True  NaN  Southampton    no  
...      ...         ...  ...          ...   ...  
True     man        True  NaN  Southampton    no  
True   woman       False    B  Southampton   yes  
False  woman       False  NaN  Southampton    no  
True     man        True    C    Cherbourg   yes  
True     man        True  NaN   Queenstown    no  

[891 rows x 14 columns]

### 8 Reset the index

In [58]:
df.reset_index(inplace = True)
df

alone  survived  pclass  Gender   age  sibsp  parch     fare embarked  \
0    False         0       3    male  22.0      1      0   7.2500        S   
1    False         1       1  female  38.0      1      0  71.2833        C   
2     True         1       3  female  26.0      0      0   7.9250        S   
3    False         1       1  female  35.0      1      0  53.1000        S   
4     True         0       3    male  35.0      0      0   8.0500        S   
..     ...       ...     ...     ...   ...    ...    ...      ...      ...   
886   True         0       2    male  27.0      0      0  13.0000        S   
887   True         1       1  female  19.0      0      0  30.0000        S   
888  False         0       3  female   NaN      1      2  23.4500        S   
889   True         1       1    male  26.0      0      0  30.0000        C   
890   True         0       3    male  32.0      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  
0     Third    man        True  NaN  Southampton    no  
1     First  woman       False    C    Cherbourg   yes  
2     Third  woman       False  NaN  Southampton   yes  
3     First  woman       False    C  Southampton   yes  
4     Third    man        True  NaN  Southampton    no  
..      ...    ...         ...  ...          ...   ...  
886  Second    man        True  NaN  Southampton    no  
887   First  woman       False    B  Southampton   yes  
888   Third  woman       False  NaN  Southampton    no  
889   First    man        True    C    Cherbourg   yes  
890   Third    man        True  NaN   Queenstown    no  

[891 rows x 15 columns]

### 9.selecting row with index position

In [74]:
df['embarked'][0:]

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: embarked, Length: 891, dtype: object

### 10.selecting subsection of the dataset with LOC and iloc

In [77]:
df.loc[887]

alone                 True
survived                 1
pclass                   1
Gender              female
age                   19.0
sibsp                    0
parch                    0
fare                  30.0
embarked                 S
class                First
who                  woman
adult_male           False
deck                     B
embark_town    Southampton
alive                  yes
Name: 887, dtype: object

In [81]:
df.loc[887::,["alone","survived","pclass","Gender","age","sibsp","parch","fare","embarked"]]

alone  survived  pclass  Gender   age  sibsp  parch   fare embarked
887   True         1       1  female  19.0      0      0  30.00        S
888  False         0       3  female   NaN      1      2  23.45        S
889   True         1       1    male  26.0      0      0  30.00        C
890   True         0       3    male  32.0      0      0   7.75        Q

In [83]:
df.iloc[887::,4:8]

age  sibsp  parch   fare
887  19.0      0      0  30.00
888   NaN      1      2  23.45
889  26.0      0      0  30.00
890  32.0      0      0   7.75

### 11.how to drop the row

In [86]:
df.drop(889,axis = 0)

alone  survived  pclass  Gender   age  sibsp  parch     fare embarked  \
0    False         0       3    male  22.0      1      0   7.2500        S   
1    False         1       1  female  38.0      1      0  71.2833        C   
2     True         1       3  female  26.0      0      0   7.9250        S   
3    False         1       1  female  35.0      1      0  53.1000        S   
4     True         0       3    male  35.0      0      0   8.0500        S   
..     ...       ...     ...     ...   ...    ...    ...      ...      ...   
885  False         0       3  female  39.0      0      5  29.1250        Q   
886   True         0       2    male  27.0      0      0  13.0000        S   
887   True         1       1  female  19.0      0      0  30.0000        S   
888  False         0       3  female   NaN      1      2  23.4500        S   
890   True         0       3    male  32.0      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  
0     Third    man        True  NaN  Southampton    no  
1     First  woman       False    C    Cherbourg   yes  
2     Third  woman       False  NaN  Southampton   yes  
3     First  woman       False    C  Southampton   yes  
4     Third    man        True  NaN  Southampton    no  
..      ...    ...         ...  ...          ...   ...  
885   Third  woman       False  NaN   Queenstown    no  
886  Second    man        True  NaN  Southampton    no  
887   First  woman       False    B  Southampton   yes  
888   Third  woman       False  NaN  Southampton    no  
890   Third    man        True  NaN   Queenstown    no  

[890 rows x 15 columns]

### 12.conditional filtering using "&" and "|"

In [105]:
### This will filter out those data where Gender is Male And Age is More the 38
df[(df['Gender']=="male") & (df['age']>38)]

survived  pclass Gender   age  sibsp  parch     fare embarked   class  \
6           0       1   male  54.0      0      0  51.8625        S   First   
13          0       3   male  39.0      1      5  31.2750        S   Third   
30          0       1   male  40.0      0      0  27.7208        C   First   
33          0       2   male  66.0      0      0  10.5000        S  Second   
35          0       1   male  42.0      1      0  52.0000        S   First   
..        ...     ...    ...   ...    ...    ...      ...      ...     ...   
845         0       3   male  42.0      0      0   7.5500        S   Third   
851         0       3   male  74.0      0      0   7.7750        S   Third   
857         1       1   male  51.0      0      0  26.5500        S   First   
860         0       3   male  41.0      2      0  14.1083        S   Third   
873         0       3   male  47.0      0      0   9.0000        S   Third   

     who  adult_male deck  embark_town alive  alone  
6    man        True    E  Southampton    no   True  
13   man        True  NaN  Southampton    no  False  
30   man        True  NaN    Cherbourg    no   True  
33   man        True  NaN  Southampton    no   True  
35   man        True  NaN  Southampton    no  False  
..   ...         ...  ...          ...   ...    ...  
845  man        True  NaN  Southampton    no   True  
851  man        True  NaN  Southampton    no   True  
857  man        True    E  Southampton   yes   True  
860  man        True  NaN  Southampton    no  False  
873  man        True  NaN  Southampton    no   True  

[117 rows x 15 columns]

In [109]:
### This Will filter out those data where Gender is Male Or Female and they are alive 
df[((df['Gender'] == "male") | (df['Gender'] == "female")) & (df['alive'] == "yes") ]

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
8           1       3  female  27.0      0      2  11.1333        S   Third   
9           1       2  female  14.0      1      0  30.0708        C  Second   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
875         1       3  female  15.0      0      0   7.2250        C   Third   
879         1       1  female  56.0      0      1  83.1583        C   First   
880         1       2  female  25.0      0      1  26.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
8    woman       False  NaN  Southampton   yes  False  
9    child       False  NaN    Cherbourg   yes  False  
..     ...         ...  ...          ...   ...    ...  
875  child       False  NaN    Cherbourg   yes   True  
879  woman       False    C    Cherbourg   yes  False  
880  woman       False  NaN  Southampton   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[342 rows x 15 columns]

In [115]:
### we also have the method of isin() You can use in for multiple condition in a single column
df[df['Gender'].isin(["male", "female"])]

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 13 apply function on single column

In [116]:
def fun(x):
    return x.title()

fun("male")

'Male'

In [119]:
### This Function Will Convert the Gender Lower Case to Proper Case 
df['Gender'] = df['Gender'].apply(fun)
df

survived  pclass  Gender   age  sibsp  parch     fare embarked   class  \
0           0       3    Male  22.0      1      0   7.2500        S   Third   
1           1       1  Female  38.0      1      0  71.2833        C   First   
2           1       3  Female  26.0      0      0   7.9250        S   Third   
3           1       1  Female  35.0      1      0  53.1000        S   First   
4           0       3    Male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    Male  27.0      0      0  13.0000        S  Second   
887         1       1  Female  19.0      0      0  30.0000        S   First   
888         0       3  Female   NaN      1      2  23.4500        S   Third   
889         1       1    Male  26.0      0      0  30.0000        C   First   
890         0       3    Male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 14 By Using Map Function We Can Also Change the Case of Columns

In [128]:
df.columns = df.columns.map(fun)
df

Survived  Pclass  Gender   Age  Sibsp  Parch     Fare Embarked   Class  \
0           0       3    Male  22.0      1      0   7.2500        S   Third   
1           1       1  Female  38.0      1      0  71.2833        C   First   
2           1       3  Female  26.0      0      0   7.9250        S   Third   
3           1       1  Female  35.0      1      0  53.1000        S   First   
4           0       3    Male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    Male  27.0      0      0  13.0000        S  Second   
887         1       1  Female  19.0      0      0  30.0000        S   First   
888         0       3  Female   NaN      1      2  23.4500        S   Third   
889         1       1    Male  26.0      0      0  30.0000        C   First   
890         0       3    Male  32.0      0      0   7.7500        Q   Third   

       Who  Adult_Male Deck  Embark_Town Alive  Alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 15 apply func on multiple column using lambda func

In [132]:
## We Have Convert the Fare float data to int
df['Fare'] = df['Fare'].apply(lambda x :int(x))
df.head()

Survived  Pclass  Gender   Age  Sibsp  Parch  Fare Embarked  Class    Who  \
0         0       3    Male  22.0      1      0     7        S  Third    man   
1         1       1  Female  38.0      1      0    71        C  First  woman   
2         1       3  Female  26.0      0      0     7        S  Third  woman   
3         1       1  Female  35.0      1      0    53        S  First  woman   
4         0       3    Male  35.0      0      0     8        S  Third    man   

   Adult_Male Deck  Embark_Town Alive  Alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

### Sorting Method

#### To sort the Dataset We can Use sort_values() Method, 1 perameter is need that is by , by which column you need to be sort the data

* Let's Say I want to Sort the Data By Fare ( Low to High)
* Then Let's Say I want to Sort the Data By Fare ( Hig to Low)

In [134]:
df.sort_values(by= "Fare")

Survived  Pclass  Gender   Age  Sibsp  Parch  Fare Embarked   Class  \
822         0       1    Male  38.0      0      0     0        S   First   
179         0       3    Male  36.0      0      0     0        S   Third   
466         0       2    Male   NaN      0      0     0        S  Second   
263         0       1    Male  40.0      0      0     0        S   First   
806         0       1    Male  39.0      0      0     0        S   First   
..        ...     ...     ...   ...    ...    ...   ...      ...     ...   
88          1       1  Female  23.0      3      2   263        S   First   
438         0       1    Male  64.0      1      4   263        S   First   
679         1       1    Male  36.0      0      1   512        C   First   
737         1       1    Male  35.0      0      0   512        C   First   
258         1       1  Female  35.0      0      0   512        C   First   

       Who  Adult_Male Deck  Embark_Town Alive  Alone  
822    man        True  NaN  Southampton    no   True  
179    man        True  NaN  Southampton    no   True  
466    man        True  NaN  Southampton    no   True  
263    man        True    B  Southampton    no   True  
806    man        True    A  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
88   woman       False    C  Southampton   yes  False  
438    man        True    C  Southampton    no  False  
679    man        True    B    Cherbourg   yes  False  
737    man        True    B    Cherbourg   yes   True  
258  woman       False  NaN    Cherbourg   yes   True  

[891 rows x 15 columns]

In [135]:
df.sort_values(by = "Fare", ascending = False)

Survived  Pclass  Gender   Age  Sibsp  Parch  Fare Embarked   Class  \
679         1       1    Male  36.0      0      1   512        C   First   
258         1       1  Female  35.0      0      0   512        C   First   
737         1       1    Male  35.0      0      0   512        C   First   
438         0       1    Male  64.0      1      4   263        S   First   
88          1       1  Female  23.0      3      2   263        S   First   
..        ...     ...     ...   ...    ...    ...   ...      ...     ...   
822         0       1    Male  38.0      0      0     0        S   First   
466         0       2    Male   NaN      0      0     0        S  Second   
271         1       3    Male  25.0      0      0     0        S   Third   
277         0       2    Male   NaN      0      0     0        S  Second   
597         0       3    Male  49.0      0      0     0        S   Third   

       Who  Adult_Male Deck  Embark_Town Alive  Alone  
679    man        True    B    Cherbourg   yes  False  
258  woman       False  NaN    Cherbourg   yes   True  
737    man        True    B    Cherbourg   yes   True  
438    man        True    C  Southampton    no  False  
88   woman       False    C  Southampton   yes  False  
..     ...         ...  ...          ...   ...    ...  
822    man        True  NaN  Southampton    no   True  
466    man        True  NaN  Southampton    no   True  
271    man        True  NaN  Southampton   yes   True  
277    man        True  NaN  Southampton    no   True  
597    man        True  NaN  Southampton    no   True  

[891 rows x 15 columns]

### 16.min , max and their index position

In [142]:
df['Fare'].min()

0

In [145]:
df['Fare'].idxmin()

179

##### the minmum Fare is 0 the their index Number is 179

In [151]:
df['Fare'].max()

512

In [152]:
df['Fare'].idxmax()

258

### 17.value_counts/unique/nunique/replace/map function

In [161]:
df.value_counts(['Gender','Class'])

Gender  Class 
Male    Third     347
Female  Third     144
Male    First     122
        Second    108
Female  First      94
        Second     76
Name: count, dtype: int64

In [167]:
df[['Gender',"Class"]].value_counts(sort = False)

Gender  Class 
Female  First      94
        Second     76
        Third     144
Male    First     122
        Second    108
        Third     347
Name: count, dtype: int64

#### 18 Unique Method

In [177]:
df['Embark_Town'].unique()

array(['Southampton', 'Cherbourg', 'Queenstown', nan], dtype=object)

### 19 nunique() Method 
This Method will return the Number of Unique Element in your Data Frame / Series

In [181]:
df.nunique()

Survived        2
Pclass          3
Gender          2
Age            88
Sibsp           7
Parch           7
Fare           91
Embarked        3
Class           3
Who             3
Adult_Male      2
Deck            7
Embark_Town     3
Alive           2
Alone           2
dtype: int64

In [189]:
df['Gender'].replace(["Female","Male"],["F","M"], inplace = True)
df

Survived  Pclass Gender   Age  Sibsp  Parch  Fare Embarked   Class  \
0           0       3      M  22.0      1      0     7        S   Third   
1           1       1      F  38.0      1      0    71        C   First   
2           1       3      F  26.0      0      0     7        S   Third   
3           1       1      F  35.0      1      0    53        S   First   
4           0       3      M  35.0      0      0     8        S   Third   
..        ...     ...    ...   ...    ...    ...   ...      ...     ...   
886         0       2      M  27.0      0      0    13        S  Second   
887         1       1      F  19.0      0      0    30        S   First   
888         0       3      F   NaN      1      2    23        S   Third   
889         1       1      M  26.0      0      0    30        C   First   
890         0       3      M  32.0      0      0     7        Q   Third   

       Who  Adult_Male Deck  Embark_Town Alive  Alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [191]:
df['Gender'].replace({"F":"Female","M":"Male"},inplace = True)

In [193]:
df.sample(10)

Survived  Pclass  Gender   Age  Sibsp  Parch  Fare Embarked   Class  \
418         0       2    Male  30.0      0      0    13        S  Second   
67          0       3    Male  19.0      0      0     8        S   Third   
614         0       3    Male  35.0      0      0     8        S   Third   
265         0       2    Male  36.0      0      0    10        S  Second   
180         0       3  Female   NaN      8      2    69        S   Third   
476         0       2    Male  34.0      1      0    21        S  Second   
781         1       1  Female  17.0      1      0    57        S   First   
330         1       3  Female   NaN      2      0    23        Q   Third   
301         1       3    Male   NaN      2      0    23        Q   Third   
272         1       2  Female  41.0      0      1    19        S  Second   

       Who  Adult_Male Deck  Embark_Town Alive  Alone  
418    man        True  NaN  Southampton    no   True  
67     man        True  NaN  Southampton    no   True  
614    man        True  NaN  Southampton    no   True  
265    man        True  NaN  Southampton    no   True  
180  woman       False  NaN  Southampton    no  False  
476    man        True  NaN  Southampton    no  False  
781  woman       False    B  Southampton   yes  False  
330  woman       False  NaN   Queenstown   yes  False  
301    man        True  NaN   Queenstown   yes  False  
272  woman       False  NaN  Southampton   yes  False

### Map Funcation

In [205]:
df['Who'].unique()

array(['man', 'woman', 'child'], dtype=object)

In [221]:
df['Who'] = df['Who'].map({"man":"M","woman":"W","child":"C"})
df.head()

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked  Class Who  \
0         0       3    male  22.0      1      0   7.2500        S  Third   M   
1         1       1  female  38.0      1      0  71.2833        C  First   W   
2         1       3  female  26.0      0      0   7.9250        S  Third   W   
3         1       1  female  35.0      1      0  53.1000        S  First   W   
4         0       3    male  35.0      0      0   8.0500        S  Third   M   

   Adult_Male Deck  Embark_Town Alive  Alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

* 19.nlargest/nsmallest to get highest/lowest records
* 20.sample of the dataset (by numbers or percentage)

### 18.treatment of duplicate values

In [225]:
df[df.duplicated()]

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked   Class  \
47          1       3  female   NaN      0      0   7.7500        Q   Third   
76          0       3    male   NaN      0      0   7.8958        S   Third   
77          0       3    male   NaN      0      0   8.0500        S   Third   
87          0       3    male   NaN      0      0   8.0500        S   Third   
95          0       3    male   NaN      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
870         0       3    male  26.0      0      0   7.8958        S   Third   
877         0       3    male  19.0      0      0   7.8958        S   Third   
878         0       3    male   NaN      0      0   7.8958        S   Third   
884         0       3    male  25.0      0      0   7.0500        S   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   

    Who  Adult_Male Deck  Embark_Town Alive  Alone  
47    W       False  NaN   Queenstown   yes   True  
76    M        True  NaN  Southampton    no   True  
77    M        True  NaN  Southampton    no   True  
87    M        True  NaN  Southampton    no   True  
95    M        True  NaN  Southampton    no   True  
..   ..         ...  ...          ...   ...    ...  
870   M        True  NaN  Southampton    no   True  
877   M        True  NaN  Southampton    no   True  
878   M        True  NaN  Southampton    no   True  
884   M        True  NaN  Southampton    no   True  
886   M        True  NaN  Southampton    no   True  

[107 rows x 15 columns]

#### Droping the Duplicated Values

In [228]:
df.drop_duplicates().head()

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked  Class Who  \
0         0       3    male  22.0      1      0   7.2500        S  Third   M   
1         1       1  female  38.0      1      0  71.2833        C  First   W   
2         1       3  female  26.0      0      0   7.9250        S  Third   W   
3         1       1  female  35.0      1      0  53.1000        S  First   W   
4         0       3    male  35.0      0      0   8.0500        S  Third   M   

   Adult_Male Deck  Embark_Town Alive  Alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

### 19.nlargest/nsmallest to get highest/lowest records

### nlargest

In [240]:
#### It'll revert those 10 maximum fare
df.nlargest(10,'Fare')

Survived  Pclass     Sex   Age  Sibsp  Parch      Fare Embarked  Class  \
258         1       1  female  35.0      0      0  512.3292        C  First   
679         1       1    male  36.0      0      1  512.3292        C  First   
737         1       1    male  35.0      0      0  512.3292        C  First   
27          0       1    male  19.0      3      2  263.0000        S  First   
88          1       1  female  23.0      3      2  263.0000        S  First   
341         1       1  female  24.0      3      2  263.0000        S  First   
438         0       1    male  64.0      1      4  263.0000        S  First   
311         1       1  female  18.0      2      2  262.3750        C  First   
742         1       1  female  21.0      2      2  262.3750        C  First   
118         0       1    male  24.0      0      1  247.5208        C  First   

    Who  Adult_Male Deck  Embark_Town Alive  Alone  
258   W       False  NaN    Cherbourg   yes   True  
679   M        True    B    Cherbourg   yes  False  
737   M        True    B    Cherbourg   yes   True  
27    M        True    C  Southampton    no  False  
88    W       False    C  Southampton   yes  False  
341   W       False    C  Southampton   yes  False  
438   M        True    C  Southampton    no  False  
311   W       False    B    Cherbourg   yes  False  
742   W       False    B    Cherbourg   yes  False  
118   M        True    B    Cherbourg    no  False

### nsmallest

In [253]:
df.nsmallest(10,'Fare')

Survived  Pclass   Sex   Age  Sibsp  Parch  Fare Embarked   Class Who  \
179         0       3  male  36.0      0      0   0.0        S   Third   M   
263         0       1  male  40.0      0      0   0.0        S   First   M   
271         1       3  male  25.0      0      0   0.0        S   Third   M   
277         0       2  male   NaN      0      0   0.0        S  Second   M   
302         0       3  male  19.0      0      0   0.0        S   Third   M   
413         0       2  male   NaN      0      0   0.0        S  Second   M   
466         0       2  male   NaN      0      0   0.0        S  Second   M   
481         0       2  male   NaN      0      0   0.0        S  Second   M   
597         0       3  male  49.0      0      0   0.0        S   Third   M   
633         0       1  male   NaN      0      0   0.0        S   First   M   

     Adult_Male Deck  Embark_Town Alive  Alone  
179        True  NaN  Southampton    no   True  
263        True    B  Southampton    no   True  
271        True  NaN  Southampton   yes   True  
277        True  NaN  Southampton    no   True  
302        True  NaN  Southampton    no   True  
413        True  NaN  Southampton    no   True  
466        True  NaN  Southampton    no   True  
481        True  NaN  Southampton    no   True  
597        True  NaN  Southampton    no   True  
633        True  NaN  Southampton    no   True

### 20.sample of the dataset (by numbers or percentage)

In [260]:
df.sample(10) ## The 10 Row of DataFrame Picked Randomly

Survived  Pclass     Sex   Age  Sibsp  Parch      Fare Embarked   Class  \
821         1       3    male  27.0      0      0    8.6625        S   Third   
450         0       2    male  36.0      1      2   27.7500        S  Second   
815         0       1    male   NaN      0      0    0.0000        S   First   
752         0       3    male  33.0      0      0    9.5000        S   Third   
554         1       3  female  22.0      0      0    7.7750        S   Third   
299         1       1  female  50.0      0      1  247.5208        C   First   
132         0       3  female  47.0      1      0   14.5000        S   Third   
776         0       3    male   NaN      0      0    7.7500        Q   Third   
525         0       3    male  40.5      0      0    7.7500        Q   Third   
314         0       2    male  43.0      1      1   26.2500        S  Second   

    Who  Adult_Male Deck  Embark_Town Alive  Alone  
821   M        True  NaN  Southampton   yes   True  
450   M        True  NaN  Southampton    no  False  
815   M        True    B  Southampton    no   True  
752   M        True  NaN  Southampton    no   True  
554   W       False  NaN  Southampton   yes   True  
299   W       False    B    Cherbourg   yes  False  
132   W       False  NaN  Southampton    no  False  
776   M        True    F   Queenstown    no   True  
525   M        True  NaN   Queenstown    no   True  
314   M        True  NaN  Southampton    no  False

In [261]:
df.sample(frac= 0.01) ### The 1% of Data Select randomly

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked   Class  \
697         1       3  female   NaN      0      0   7.7333        Q   Third   
657         0       3  female  32.0      1      1  15.5000        Q   Third   
600         1       2  female  24.0      2      1  27.0000        S  Second   
338         1       3    male  45.0      0      0   8.0500        S   Third   
107         1       3    male   NaN      0      0   7.7750        S   Third   
354         0       3    male   NaN      0      0   7.2250        C   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
597         0       3    male  49.0      0      0   0.0000        S   Third   
194         1       1  female  44.0      0      0  27.7208        C   First   

    Who  Adult_Male Deck  Embark_Town Alive  Alone  
697   W       False  NaN   Queenstown   yes   True  
657   W       False  NaN   Queenstown    no  False  
600   W       False  NaN  Southampton   yes  False  
338   M        True  NaN  Southampton   yes   True  
107   M        True  NaN  Southampton   yes   True  
354   M        True  NaN    Cherbourg    no   True  
3     W       False    C  Southampton   yes  False  
597   M        True  NaN  Southampton    no   True  
194   W       False    B    Cherbourg   yes   True

### 21.handling missing data

We can Handle the Missing Value by droping or by using the filling missing value technique. As per companies Standard if Our Data of Columns Containts More Then 33% missing Value Then We have the drop the entire columns Else We Need to fill the Missing value using some technique 

### 22 isnull() / notnull()

In [276]:
### This is percentage of Missing Value in Our Columns
(df.isnull().sum() * 100 / len(df)).round(2)

Survived        0.00
Pclass          0.00
Sex             0.00
Age            19.87
Sibsp           0.00
Parch           0.00
Fare            0.00
Embarked        0.22
Class           0.00
Who             0.00
Adult_Male      0.00
Deck           77.22
Embark_Town     0.22
Alive           0.00
Alone           0.00
dtype: float64

### Notnull

In [278]:
(df.notnull().sum() * 100 // len(df)).round(4)

Survived       100
Pclass         100
Sex            100
Age             80
Sibsp          100
Parch          100
Fare           100
Embarked        99
Class          100
Who            100
Adult_Male     100
Deck            22
Embark_Town     99
Alive          100
Alone          100
dtype: int64

### 23.dropping missing values


In [310]:
#### As We can See Our Deck Columns Contain More the 77% missing Values So We Can directly Drop the column
df.drop('Deck', axis =1 , inplace = True)
df

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked   Class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female  24.0      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       Who  Adult_Male  Embark_Town Alive  Alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

### 24.filling missing values

In [284]:
df.isnull().sum() / len(df) * 100

Survived        0.000000
Pclass          0.000000
Sex             0.000000
Age            19.865320
Sibsp           0.000000
Parch           0.000000
Fare            0.000000
Embarked        0.224467
Class           0.000000
Who             0.000000
Adult_Male      0.000000
Embark_Town     0.224467
Alive           0.000000
Alone           0.000000
dtype: float64

In [289]:
### We Can Missing the missing age values using Mean of Age
mean_age = df['Age'].mean()
print(mean_age)

df['Age'].fillna(mean_age).isnull().sum()

### As We Can See Now We Have 0 Missing Value in Age Column But We Haven't Use the inplace , perameter because we are going to 
### use another method of handling the missing values in Age Column

29.69911764705882


0

In [307]:
df['Age'].fillna(df['Age'].mode().iloc[0],inplace = True)

In [308]:
df['Age'].isnull().sum()

0

### 25.group by operation on pandas

In [311]:
df.head()

Survived  Pclass     Sex   Age  Sibsp  Parch     Fare Embarked  Class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     Who  Adult_Male  Embark_Town Alive  Alone  
0    man        True  Southampton    no  False  
1  woman       False    Cherbourg   yes  False  
2  woman       False  Southampton   yes   True  
3  woman       False  Southampton   yes  False  
4    man        True  Southampton    no   True

In [330]:
### This Group By Record will show how many people are alive and dead gender wise in Titanic
df.groupby(['Sex','Alive'])['Alive'].count()

Sex     Alive
female  no        81
        yes      233
male    no       468
        yes      109
Name: Alive, dtype: int64

In [362]:
new_df = sns.load_dataset("tips")
temp_df =  new_df[new_df.columns[(new_df.dtypes == 'float64')]]
temp_df.groupby(['total_bill','tip']).mean()

Empty DataFrame
Columns: []
Index: [(3.07, 1.0), (5.75, 1.0), (7.25, 1.0), (7.25, 5.15), (7.51, 2.0), (7.56, 1.44), (7.74, 1.44), (8.35, 1.5), (8.51, 1.25), (8.52, 1.48), (8.58, 1.92), (8.77, 2.0), (9.55, 1.45), (9.6, 4.0), (9.68, 1.32), (9.78, 1.73), (9.94, 1.56), (10.07, 1.25), (10.07, 1.83), (10.09, 2.0), (10.27, 1.71), (10.29, 2.6), (10.33, 1.67), (10.33, 2.0), (10.34, 1.66), (10.34, 2.0), (10.51, 1.25), (10.59, 1.61), (10.63, 2.0), (10.65, 1.5), (10.77, 1.47), (11.02, 1.98), (11.17, 1.5), (11.24, 1.76), (11.35, 2.5), (11.38, 2.0), (11.59, 1.5), (11.61, 3.39), (11.69, 2.31), (11.87, 1.63), (12.02, 1.97), (12.03, 1.5), (12.16, 2.2), (12.26, 2.0), (12.43, 1.8), (12.46, 1.5), (12.48, 2.52), (12.54, 2.5), (12.6, 1.0), (12.66, 2.5), (12.69, 2.0), (12.74, 2.01), (12.76, 2.23), (12.9, 1.1), (13.0, 2.0), (13.03, 2.0), (13.13, 2.0), (13.16, 2.75), (13.27, 2.5), (13.28, 2.72), (13.37, 2.0), (13.39, 2.61), (13.42, 1.58), (13.42, 1.68), (13.42, 3.48), (13.51, 2.0), (13.81, 2.0), (13.94, 3.06), (14.0, 3.0), (14.07, 2.5), (14.15, 2.0), (14.26, 2.5), (14.31, 4.0), (14.48, 2.0), (14.52, 2.0), (14.73, 2.2), (14.78, 3.23), (14.83, 3.02), (15.01, 2.09), (15.04, 1.96), (15.06, 3.0), (15.36, 1.64), (15.38, 3.0), (15.42, 1.57), (15.48, 2.02), (15.53, 3.0), (15.69, 1.5), (15.69, 3.0), (15.77, 2.23), (15.81, 3.16), (15.95, 2.0), (15.98, 2.03), (15.98, 3.0), (16.0, 2.0), (16.04, 2.24), (16.21, 2.0), (16.27, 2.5), (16.29, 3.71), (16.31, 2.0), (16.32, 4.3), ...]

[242 rows x 0 columns]

### 26.combining dataframe-->concatenation

In [363]:
data_one={"A":["A0","A1","A2","A3"],"B":["B0","B1","B2","B3"]}
data_Two={"C":["C0","C1","C2","C3"],"D":["D0","D1","D2","D3"]}

In [365]:
df1 = pd.DataFrame(data_one)
df1

A   B
0  A0  B0
1  A1  B1
2  A2  B2
3  A3  B3

In [368]:
df2 = pd.DataFrame(data_Two)
df2

C   D
0  C0  D0
1  C1  D1
2  C2  D2
3  C3  D3

In [371]:
df = pd.concat([df1,df2], axis =1)
df

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

### 27.combining dataframe--> merging-->join--> inner/left/right/outer

In [381]:
students = pd.DataFrame({'ID':[1,2,3,4],'name':['Andrew','Bobo','Claire','David'],'course_id':[3,2,4,1]})
course = pd.DataFrame({'course_id':[1,2,3,4],'course':['Excel','Python','Mysql','Java']})

In [383]:
students

ID    name  course_id
0   1  Andrew          3
1   2    Bobo          2
2   3  Claire          4
3   4   David          1

In [385]:
course

course_id  course
0          1   Excel
1          2  Python
2          3   Mysql
3          4    Java

### Join
* Join is used to join the Data in Sql as same we can use it in pandas as well
* There Are 5 types of join Inner , Outer , left , right , cross , inner

### Inner Join

In [396]:
pd.merge(students,course, how = 'inner', on = 'course_id')

ID    name  course_id  course
0   1  Andrew          3   Mysql
1   2    Bobo          2  Python
2   3  Claire          4    Java
3   4   David          1   Excel

### Left Outer Join

In [401]:
course['course_id'].replace({4:6},inplace = True)

In [403]:
pd.merge(students, course, how = 'left', on = 'course_id')

ID    name  course_id  course
0   1  Andrew          3   Mysql
1   2    Bobo          2  Python
2   3  Claire          4     NaN
3   4   David          1   Excel

### Right Outer Join

In [405]:
pd.merge(students, course, how = 'right', on = 'course_id')

ID    name  course_id  course
0  4.0   David          1   Excel
1  2.0    Bobo          2  Python
2  1.0  Andrew          3   Mysql
3  NaN     NaN          6    Java

### Cross Join

In [413]:
pd.merge(students, course, how = 'cross')

ID    name  course_id_x  course_id_y  course
0    1  Andrew            3            1   Excel
1    1  Andrew            3            2  Python
2    1  Andrew            3            3   Mysql
3    1  Andrew            3            6    Java
4    2    Bobo            2            1   Excel
5    2    Bobo            2            2  Python
6    2    Bobo            2            3   Mysql
7    2    Bobo            2            6    Java
8    3  Claire            4            1   Excel
9    3  Claire            4            2  Python
10   3  Claire            4            3   Mysql
11   3  Claire            4            6    Java
12   4   David            1            1   Excel
13   4   David            1            2  Python
14   4   David            1            3   Mysql
15   4   David            1            6    Java

### 28.text method on string data

In [417]:
data = pd.read_csv("C:\\Users\\AnshSrivastav\\Desktop\\TEMP\\Test_Query_2023-12-05.csv")
data = data[data.columns[2:4]]
data.head()

Subject                                        Description
0  Test Question Issue  Batch Name : Arjuna JEE 2024 Test : Practice T...
1  Test Question Issue  Batch Name : Arjuna JEE 2024 Test : Short Prac...
2  Test Question Issue  Test Series : Test : 10th_Foundation JEE_C_2 S...
3  Test Question Issue  Batch Name : Lakshya JEE 2024 Test : Advance T...
4  Test Question Issue  Batch Name : Arjuna JEE 2024 Test : Practice T...

In [419]:
data['Description'].str.lower()

0      batch name : arjuna jee 2024 test : practice t...
1      batch name : arjuna jee 2024 test : short prac...
2      test series : test : 10th_foundation jee_c_2 s...
3      batch name : lakshya jee 2024 test : advance t...
4      batch name : arjuna jee 2024 test : practice t...
                             ...                        
217    batch name : test : practice test 04 section :...
218    batch name : lakshya jee 2024 test : practice ...
219    test series : test : 11th_jee_c_2 section : ph...
220    batch name : fast track shaurya nda crash cour...
221    batch name : test : practice test 09 (legend) ...
Name: Description, Length: 222, dtype: object

In [420]:
data['Description'].str.upper()

0      BATCH NAME : ARJUNA JEE 2024 TEST : PRACTICE T...
1      BATCH NAME : ARJUNA JEE 2024 TEST : SHORT PRAC...
2      TEST SERIES : TEST : 10TH_FOUNDATION JEE_C_2 S...
3      BATCH NAME : LAKSHYA JEE 2024 TEST : ADVANCE T...
4      BATCH NAME : ARJUNA JEE 2024 TEST : PRACTICE T...
                             ...                        
217    BATCH NAME : TEST : PRACTICE TEST 04 SECTION :...
218    BATCH NAME : LAKSHYA JEE 2024 TEST : PRACTICE ...
219    TEST SERIES : TEST : 11TH_JEE_C_2 SECTION : PH...
220    BATCH NAME : FAST TRACK SHAURYA NDA CRASH COUR...
221    BATCH NAME : TEST : PRACTICE TEST 09 (LEGEND) ...
Name: Description, Length: 222, dtype: object

In [421]:
data['Description'].str.title()

0      Batch Name : Arjuna Jee 2024 Test : Practice T...
1      Batch Name : Arjuna Jee 2024 Test : Short Prac...
2      Test Series : Test : 10Th_Foundation Jee_C_2 S...
3      Batch Name : Lakshya Jee 2024 Test : Advance T...
4      Batch Name : Arjuna Jee 2024 Test : Practice T...
                             ...                        
217    Batch Name : Test : Practice Test 04 Section :...
218    Batch Name : Lakshya Jee 2024 Test : Practice ...
219    Test Series : Test : 11Th_Jee_C_2 Section : Ph...
220    Batch Name : Fast Track Shaurya Nda Crash Cour...
221    Batch Name : Test : Practice Test 09 (Legend) ...
Name: Description, Length: 222, dtype: object

In [426]:
data['Description'].str.capitalize()

0      Batch name : arjuna jee 2024 test : practice t...
1      Batch name : arjuna jee 2024 test : short prac...
2      Test series : test : 10th_foundation jee_c_2 s...
3      Batch name : lakshya jee 2024 test : advance t...
4      Batch name : arjuna jee 2024 test : practice t...
                             ...                        
217    Batch name : test : practice test 04 section :...
218    Batch name : lakshya jee 2024 test : practice ...
219    Test series : test : 11th_jee_c_2 section : ph...
220    Batch name : fast track shaurya nda crash cour...
221    Batch name : test : practice test 09 (legend) ...
Name: Description, Length: 222, dtype: object

In [428]:
df = pd.DataFrame([("Ansh ",1,"Male "),("Kumar ",0,"Male") , ("Sheela ",1,"Female")], columns = ['Name','Is_Alive','Gender'])
df

Name  Is_Alive  Gender
0    Ansh          1   Male 
1   Kumar          0    Male
2  Sheela          1  Female

### .cleaning the data

In [430]:
### As We Can See there Having Some Extra White space in DataFrame We Need to Remove
df['Name'][0]

'Ansh '

In [434]:
df['Name'].str.strip()[0]

'Ansh'

In [435]:
df['Mail'] = ["ansh.srivastav@outlook.com","kumar1@gmail.com","sheela.gupta@hotmail.com"]
df

Name  Is_Alive  Gender                        Mail
0    Ansh          1   Male   ansh.srivastav@outlook.com
1   Kumar          0    Male            kumar1@gmail.com
2  Sheela          1  Female    sheela.gupta@hotmail.com

### let's say we want to set user name by extracting the name from the mail and put some random number 

In [480]:
import random as nr
df['User_Name'] = df["Mail"].map(lambda x : x[0:x.find("@")] + str(nr.randint(0,11111)))

In [482]:
df

Name  Is_Alive  Gender                        Mail           User_Name
0    Ansh          1   Male   ansh.srivastav@outlook.com  ansh.srivastav8619
1   Kumar          0    Male            kumar1@gmail.com           kumar1854
2  Sheela          1  Female    sheela.gupta@hotmail.com    sheela.gupta2682

<div : class = "alert-danger alert"> 
    Thank You 